File: archives/C#/0152. Maximum Product Subarray.cs
public class Solution {
    public int MaxProduct(int[] nums) {
        if(nums.Length==0)
            return 0;
        int curMax=1;
        int curMin=1;
        int maxProduct=nums[0];
        foreach(int num in nums){
            int preMax=curMax;
            curMax=Math.Max(num,Math.Max(num*preMax,num*curMin));
            curMin=Math.Min(num,Math.Min(num*preMax,num*curMin));
            maxProduct=Math.Max(maxProduct,curMax);
            //Console.WriteLine("{0},{1},{2}",curMax,curMin,maxProduct);
        }
        return maxProduct;
    }
}

File: archives/20190519python/0152. Maximum Product Subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        rep=nums[0]
        min_nums=nums[0]
        max_nums=nums[0]
        for num in nums[1:]:
            min_nums,max_nums=min(min_nums*num,num,max_nums*num),max(min_nums*num,num,max_nums*num)
            rep=max(rep,max_nums)
        return rep
